Always allow "no-other-choice" pre-release dependencies when resolving collection dependency tree

webknjaz opened the pull request:
Prior to this patch, when `--pre` CLI flag was not passed, the
dependency resolver would treat concrete collection dependency
candidates (Git repositories, subdirs, tarball URLs, or local dirs or
files etc) as not meeting the requirements.

This patch makes it so pre-releases in any concrete artifact
references, and the ones being specifically pinned dependencies or
user requests, met anywhere in the dependency tree, are allowed
unconditionally.

This is achieved by moving the pre-release check from
`is_satisfied_by()` to the `find_matches()` hook, following the
Pip's example.

As a bonus, this change also fixes the situation when a collection
pre-releases weren't considered if it didn't have any stable releases.
This now works even if `--pre` wasn't requested explicitly.

Ref #73416
Ref #79112
Fixes #79168
Fixes #80048
Resolves #81605

Co-Authored-By: Sloane Hertel <19572925+s-hertel@users.noreply.github.com>

##### SUMMARY

<!--- Describe the change below, including rationale and design decisions -->
$sbj.

##### ISSUE TYPE

<!--- Pick one below and delete the rest -->

- Bugfix Pull Request

##### ADDITIONAL INFORMATION

Special thanks to @pradyunsg for hinting how to process pre-releases in the resolvelib provider correctly.

webknjaz committed:
Always allow pinned pre-release dependencies

Prior to this patch, when `--pre` CLI flag was not passed, the
dependency resolver would treat concrete collection dependency
candidates (Git repositories, subdirs, tarball URLs, or local dirs or
files etc) as not meeting the requirements.

This patch makes it so pre-releases in any concrete artifact
references, and the ones being specifically pinned dependencies or
user requests, met anywhwere in the dependency tree, are allowed
unconditionally.

This is achieved by moving the pre-release check from
`is_satisfied_by()` to the `find_matches()` hook, following the
Pip's example.

Ref #79112
Fixes #79168
Resolves #81605

Co-Authored-By: Sloane Hertel <19572925+s-hertel@users.noreply.github.com>

webknjaz commented:
@s-hertel I borrowed your tests from #81605 but haven't added any extra ones. I think more might be needed.

s-hertel commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -411,12 +411,18 @@ def is_satisfied_by(self, requirement, candidate):
+        # NOTE: pre-releases are allowed regardless of any other circumstances.
+        allow_pre_release = (
+            self._with_pre_releases or
+            candidate.is_concrete_artifact or not (
```
I think this is okay, though it is much broader than `type: git`, and we don't have much test coverage for dependencies that are concrete artifacts as it pertains to pre-releases. At least we know the missing test coverage/potential issues of the change is limited to dependencies, since direct reqs were (mostly) handled by the last `or`.

If this does have any unintended side effects, the only workaround would probably be `--no-deps`.

webknjaz replied:
I meant for this to apply beyond Git as while making this patch, I recognized that the problem would manifest in each of those cases otherwise. So I've gone for a generic solution. But you're right, we should have better coverage.

s-hertel replied:
I was questioning the additional change on top of that regarding dependencies (as the lay-user will see them). The generic fix is a bit redundant, but once the dead code is removed the simplification definitely seems like a net positive if it doesn't have side effects.

webknjaz replied:
Reverted. Marking this as resolved.

webknjaz replied:
I don't think it's dead. It just covers a different part of the corner case.

webknjaz replied:
Alright.. After more testing and tweaks, I mark this thread as resolved.

s-hertel commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -411,12 +411,18 @@ def is_satisfied_by(self, requirement, candidate):
+                requirement.ver.startswith('<') or
+                requirement.ver.startswith('>') or
+                requirement.ver.startswith('!=')
+            ) or self._is_user_requested(candidate)
```
I think we should clean up the (now) unused stuff that was added in https://github.com/ansible/ansible/pull/73416/files, and rename `_is_user_requested` method to something more accurate since it's now just handling a `type: galaxy` corner case.

webknjaz replied:
Yeah, maybe. I didn't look that far, just wanted to demonstrate the minimum generic change necessary.

s-hertel replied:
It's a small change, but I am less confident about backwards compatibility.

webknjaz replied:
@s-hertel looks like that entire commit can be reverted :thinking:

webknjaz replied:
Reverted. Marking this as resolved.

webknjaz replied:
Had to revert it back. It seems to still be necessary.

webknjaz replied:
Thinking more, we might need that `is_pinned` property after all. To facilitate removal of `_is_user_requested`.

webknjaz replied:
> @s-hertel looks like that entire commit can be reverted 🤔

UPD: So while working on this, just reverting the commit didn't work, which sent me off into the rabbit hole. I realized that we have a deeper problem with how we process pre-releases — if a requirement on any level of depth in the dependency tree has a pinned dependency that wan't taken into account. With the help of my pip maintainer  friend, we've confirmed that pip does not have the same problem — if a pinned dependency is a pre-release, it's allowed automatically in pip, without requiring a `--pre` which I regard as the correct behavior. With that, _I knew_ what bug we had but couldn't make up what's the difference with how pip does things. Luckily, Pradyun pointed out the pip's debugging reporter and I used it to figure out what the resolver does. He also made a comment along the lines of processing pre-releases in a different hook — `find_matches()`, not `is_satisfied_by()`! This all led took me to the right exploration path and I think I've finally figured it out: _we need to stop disregarding pinned versions from any requirements on the current round that are pinned!_

It looks like I nailed it, finally. This PR needs some love before getting out of Draft but it feels close to being ready.

@s-hertel feel free to run some testing of your own.

webknjaz replied:
This and more corner cases have been addressed. Marking this thread as resolved.

s-hertel reviewed (commented):
Does the pre-release hint https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/collection/__init__.py#L1817 need a qualification since `--pre` will have no bearing on concrete artifacts (direct or indirect)?

webknjaz commented:
> Does the pre-release hint [`devel`/lib/ansible/galaxy/collection/__init__.py#L1817](https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/collection/__init__.py?rgh-link-date=2023-08-31T13%3A39%3A32Z#L1817) need a qualification since `--pre` will have no bearing on concrete artifacts (direct or indirect)?

I'm not sure what you're suggesting. Do mean that we should change the message to say “except for concrete artifacts”? Could you elaborate?

webknjaz committed:
Revert "Match user-requested transitive pre-releases in collection dependency resolver (#73416)"

This reverts commit 6f4b4c345b44d84cce55ba6502a72b420a52c7b3.

Partially, except for the tests.

webknjaz commented:
> how about saying it it only applies to requirements from galaxy servers, since that would be more meaningful to a user.

That's a great idea!

webknjaz committed:
Hint the `--pre` affects Galaxy collections

webknjaz commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -339,15 +292,45 @@ def _find_matches(self, requirements):
                 for version, _none_src_server in coll_versions
             ]
 
+        def requirement_is_pinned(requirement):
```
_TODO:_ Move this to a requirement method.

webknjaz replied:
Moved. Marking as resolved.

webknjaz commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -339,15 +292,45 @@ def _find_matches(self, requirements):
+                    # NOTE: Only allow pre-release candidates if we want pre-releases
+                    # NOTE: or the req ver was an exact match with the pre-release
+                    # NOTE: version. Another case where we'd want to allow
+                    # NOTE: pre-releases is when there are several user requirements
```
_TODO:_ This doesn't need clarification anymore as it's now more generic.

webknjaz replied:
Updated. Marking as resolved.

webknjaz commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -339,15 +292,45 @@ def _find_matches(self, requirements):
+
         latest_matches = []
         signatures = []
         extra_signature_sources = []  # type: list[str]
```
_TODO:_ Figure out why this isn't reset on each outer loop cycle below.

webknjaz replied:
@s-hertel any ideas?

webknjaz replied:
@s-hertel this seems like a bug but I chose not to attempt addressing it within this PR. Still, it'd be good to figure it out...

webknjaz commented on test/integration/targets/ansible-galaxy-collection/tasks/main.yml:
```diff
@@ -134,6 +134,10 @@
   loop_control:
     loop_var: resolvelib_version
 
+- name: test installing prereleases via scm direct requests
```
_TODO:_ Include a dependency tree truly nested transitive pinned dependencies.

webknjaz replied:
Done. Marking as resolved.

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -1817,15 +1817,15 @@ def _resolve_depenency_map(
-        pre_release_hint = 'Hint: Pre-releases are not installed by default unless the specific version is given. To enable pre-releases, use --pre.'
+    pre_release_hint = '' if allow_pre_release else (
+        'Hint: Pre-releases hosted on Galaxy or Automation Hub are not '
+        'installed by default unless the version is given. To enable '
```
_FIXME:_ "specific" got lost

webknjaz replied:
Fixed. Marking as resolved.

webknjaz committed:
Use pre-release if collection doesn't have stable

This patch is aimed at allowing the dependency resolver to actually
consider pre-release versions of collections that don't have any
stable releases published without `--pre` being requested explicitly.

Fixes #80048

webknjaz commented:
@s-hertel I'm happy with how the patch turned out, it's now ready to be reviewed. If you have more test ideas, please send them my way.

webknjaz committed:
Add a change note for PR #81606

jborean93 commented on changelogs/fragments/81606-ansible-galaxy-collection-pre-releases.yml:
```diff
@@ -0,0 +1,16 @@
+  collections and not version ranges.
+  (https://github.com/ansible/ansible/pull/81606)
+- >-
+  ansible-galaxy - started allowing the use of pre-releases
```
Does pip work like this?

webknjaz replied:
Yep

webknjaz replied:
Retrieved some context for history:
* pip's rationale (which I'd apparently thumbed up years ago): https://github.com/pypa/pip/issues/5503#issuecomment-397426044
* PEP 440: https://peps.python.org/pep-0440/#handling-of-pre-releases

While googling I've also found that another tool — pipenv — might have a problem with this, but not pip itself:
* https://github.com/pypa/pipenv/issues/2022#issuecomment-382706536
* https://github.com/psf/black/issues/2115

jborean93 commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -564,6 +564,15 @@ def is_concrete_artifact(self):
 
+    @property
+    def is_pinned(self):
+        """Indicate if the version set is considered pinned."""
```
Might be good to expand on what `pinned` actually means in this context.

webknjaz replied:
Great suggestion, will do!

webknjaz replied:
Addressed in https://github.com/ansible/ansible/pull/81606/commits/e683693fd6a2eea1ce6283dd0b7613316ff74e32.

Marking the thread as resolved.

jborean93 reviewed (commented):
Just some minor comments, the rest looks good to me.

webknjaz committed:
Add a docstring to `Requirement.is_pinned`

webknjaz committed:
Skip computing `is_pinned` when unnecessary

This patch optimizes the case when a requirement to install a
collection is met and said collection only has pre-release versions
published with no stable versions available.

webknjaz commented:
Note for context: @s-hertel and I agreed in a private discussion on Slack, that any further improvements would go into follow-up PRs.

webknjaz commented:
Backports PRs:
* https://github.com/ansible/ansible/pull/81746
* https://github.com/ansible/ansible/pull/81747
* https://github.com/ansible/ansible/pull/81748
* https://github.com/ansible/ansible/pull/81749
